cpp/oneapi/dal/backend/primitives/objective_function/test/logloss_dpc.cpp

```diff
@@ -0,0 +1,498 @@
+        }
+    }
+
+    void test_formula_derivative(const ndarray<float_t, 2>& data,
+                                 const ndarray<float_t, 1>& probabilities,
+                                 const ndarray<float_t, 1>& params,
+                                 const ndarray<std::int32_t, 1>& labels,
+                                 const ndarray<float_t, 1>& derivative,
+                                 const float_t L1,
+                                 const float_t L2,
+                                 const float_t rtol = 1e-3,
+                                 const float_t atol = 1e-3) {
+        const std::int64_t p = data.get_dimension(1);
+        auto out_derivative =
+            ndarray<long double, 1>::empty(this->get_queue(), { p + 1 }, sycl::usm::alloc::host);
+
+        naive_derivative(data, probabilities, params, labels, out_derivative, L1, L2);
+
+        for (std::int64_t i = 0; i < p + 1; ++i) {
+            check_val(out_derivative.at(i), derivative.at(i), rtol, atol);
+        }
+    }
+
+    void test_formula_hessian(const ndarray<float_t, 2>& data,
+                              const ndarray<float_t, 1>& probabilities,
+                              const ndarray<float_t, 2>& hessian,
+                              const float_t L2,
+                              const float_t rtol = 1e-3,
+                              const float_t atol = 1e-3) {
+        const std::int64_t p = data.get_dimension(1);
+        auto out_hessian = ndarray<long double, 2>::empty(this->get_queue(),
+                                                          { p + 1, p + 1 },
+                                                          sycl::usm::alloc::host);
+
+        naive_hessian(data, probabilities, out_hessian, L2);
+
+        for (std::int64_t i = 0; i <= p; ++i) {
+            for (std::int64_t j = 0; j <= p; ++j) {
+                check_val(out_hessian.at(i, j), hessian.at(i, j), rtol, atol);
+            }
+        }
+    }
+
+    void test_derivative_and_hessian(const ndarray<float_t, 2>& data,
+                                     const ndarray<std::int32_t, 1>& labels,
+                                     const ndarray<float_t, 1>& derivative,
+                                     const ndarray<float_t, 2>& hessian,
+                                     const ndarray<float_t, 1>& params_host,
+                                     const float_t L1,
+                                     const float_t L2,
+                                     const float_t rtol = 1e-3,
+                                     const float_t atol = 1e-3) {
+        const std::int64_t n = data.get_dimension(0);
+        const std::int64_t p = data.get_dimension(1);
+        constexpr std::int64_t MAXN = 2000;
+        constexpr float_t step = sizeof(float_t) > 4 ? 1e-4 : 1e-3;
+
+        const auto data_host = data.to_host(this->get_queue());
+        const auto labels_host = labels.to_host(this->get_queue());
+
+        float_t cur_param[MAXN];
```

Comment:
Please use safer alternative:

```suggestion
        std::array<float_t, max_n> cur_param;
```